File: kube/src/main/java/com/jacksonrakena/infrastructure/envs/prod/ProductionCredentials.kt
package com.jacksonrakena.infrastructure.envs.prod

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import org.cdk8s.Chart
import org.cdk8s.ChartProps
import org.cdk8s.plus28.ConfigMap
import org.cdk8s.plus28.ConfigMapProps
import org.cdk8s.plus28.DockerConfigSecret
import org.cdk8s.plus28.DockerConfigSecretProps
import org.cdk8s.plus28.IConfigMap
import org.cdk8s.plus28.Secret
import org.cdk8s.plus28.SecretProps
import software.constructs.Construct
import java.nio.file.Files
import java.nio.file.Path

class ProductionCredentials(scope: Construct, id: String, options: ChartProps) : Chart(scope, id, options) {
    fun createNamespacedConfigMapFromFile(
        id: String,
        fileName: String
    ): IConfigMap {
        return ConfigMap(
            this,
            id,
            ConfigMapProps.builder()
                .immutable(false)
                .data(createStringDataFromFile(fileName))
                .build()
        )
    }

    fun createNamespacedSecretFromFile(
        id: String,
        fileName: String
    ): Secret {
        return Secret(
            this,
            id,
            SecretProps.builder()
                .stringData(
                    createStringDataFromFile(fileName)
                ).immutable(false).build()
        )
    }

    fun createStringDataFromFile(name: String): Map<String, String> {
        return Files.readAllLines(Path.of(name)).map { line ->
            val out = line.split("=")
            return@map Pair(out[0], out[1])
        }.toMap()
    }

    val vaultwardenSecret = createNamespacedSecretFromFile("vault-secret", "secrets/vaultwarden.env")
    val postgresSecret = createNamespacedSecretFromFile(
        "pg-secret",
        "secrets/galahad-secret.env"
    )

    val githubRegistrySecret = DockerConfigSecret(
        this, "gh-secret", DockerConfigSecretProps.builder()
            .immutable(false)
            .data(
                ObjectMapper().readValue(
                    Files.readString(Path.of("secrets/docker-registry-config.json")),
                    object : TypeReference<HashMap<String, Object>>() {}
                )
            )
            .build())

    val gradekeeperConfigMap = createNamespacedConfigMapFromFile(
        "gradekeeper-config",
        "secrets/gradekeeper-server.env"
    )

    val keycloakConfigMap = createNamespacedConfigMapFromFile(
        "keycloak-config",
        "secrets/keycloak.env"
    )

    val jacksonbotConfigMap = ConfigMap(
        this,
        "jacksonbot-config",
        ConfigMapProps.builder()
            .data(
                mapOf(
                    "jacksonbot.appsettings.json" to Files.readString(Path.of("secrets/jacksonbot-config.json"))
                )
            )
            .immutable(false).build()
    )

    val mixerConfigMap = createNamespacedConfigMapFromFile(
        "mixer-config",
        "secrets/mixer.env"
    )

    val mxbudgetConfigMap = createNamespacedConfigMapFromFile(
        "mxbudget-config",
        "secrets/mxbudget.env"
    )
}

File: src/envs/prod/production-credentials.ts
import { Construct } from "constructs";
import { Chart, ChartProps } from "cdk8s";
import * as kplus from "cdk8s-plus-34";
import * as fs from "fs";

function createStringDataFromFile(fileName: string): Record<string, string> {
  const lines = fs.readFileSync(fileName, "utf-8").split("\n").filter(Boolean);
  const data: Record<string, string> = {};
  for (const line of lines) {
    const trimmed = line.replace(/\r$/, "");
    const idx = trimmed.indexOf("=");
    if (idx > 0) {
      data[trimmed.substring(0, idx)] = trimmed.substring(idx + 1);
    }
  }
  return data;
}

export class ProductionCredentials extends Chart {
  public readonly vaultwardenSecret: kplus.Secret;
  public readonly githubRegistrySecret: kplus.DockerConfigSecret;
  public readonly gradekeeperConfigMap: kplus.IConfigMap;
  public readonly mixerBackendConfigMap: kplus.IConfigMap;

  constructor(scope: Construct, id: string, props: ChartProps) {
    super(scope, id, props);

    this.vaultwardenSecret = new kplus.Secret(this, "vault-secret", {
      stringData: createStringDataFromFile("secrets/vaultwarden.env"),
      immutable: false,
    });

    const dockerConfig = JSON.parse(
      fs.readFileSync("secrets/docker-registry-config.json", "utf-8"),
    );
    this.githubRegistrySecret = new kplus.DockerConfigSecret(
      this,
      "gh-secret",
      {
        immutable: false,
        data: dockerConfig,
      },
    );

    this.gradekeeperConfigMap = new kplus.ConfigMap(
      this,
      "gradekeeper-config",
      {
        immutable: false,
        data: createStringDataFromFile("secrets/gradekeeper-server.env"),
      },
    );

    this.mixerBackendConfigMap = new kplus.ConfigMap(
      this,
      "mixer-backend-config",
      {
        immutable: false,
        data: createStringDataFromFile("secrets/mixer.backend.env"),
      },
    );
  }
}
